File: example-moose-projects/live-hr-leaderboard-python/moose-backend/app/functions/UserActivity__ParsedActivity.py
from moose_lib import StreamingFunction
from datetime import datetime
from app.datamodels.models import UserActivity, ParsedActivity

def parse_activity(activity: UserActivity) -> ParsedActivity:
    return ParsedActivity(
        eventId=activity.eventId,
        timestamp=datetime.fromisoformat(activity.timestamp),
        userId=activity.userId,
        activity=activity.activity,
    )

my_function = StreamingFunction(
    run=parse_activity
)

File: templates/brainwaves/apps/brainmoose/app/functions/UserActivity__ParsedActivity.ts
// DMV2 Streaming Function: Converts local timestamps in UserActivity data to UTC.
// Using DMV2 IngestPipeline streams for type-safe stream processing

import {
  ParsedActivity,
  UserActivity,
  UserActivityPipeline,
  ParsedActivityPipeline,
} from "../datamodels/models";

// DMV2 Stream Transformation using pipeline streams
UserActivityPipeline.stream!.addTransform(
  ParsedActivityPipeline.stream!,
  (source: UserActivity): ParsedActivity => {
    // Convert local timestamp to UTC and return new ParsedActivity object.
    return {
      eventId: source.eventId, // Retain original event ID.
      userId: "puid" + source.userId, // Example: Prefix user ID.
      activity: source.activity, // Copy activity unchanged.
      timestamp: new Date(source.timestamp), // Convert timestamp to UTC.
    };
  },
  {
    metadata: { description: "Convert UserActivity timestamps to UTC" },
  },
);
